arrow-fx-reaktive/src/main/kotlin/arrow/fx/reaktive/MaybeK.kt

```diff
@@ -0,0 +1,330 @@
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+typealias MaybeKProc<A> = ((Either<Throwable, A>) -> Unit) -> Unit
+typealias MaybeKProcF<A> = ((Either<Throwable, A>) -> Unit) -> Kind<ForMaybeK, Unit>
+
+class ForMaybeK private constructor() {
+  companion object
+}
+typealias MaybeKOf<A> = arrow.Kind<ForMaybeK, A>
+
+@Suppress("UNCHECKED_CAST", "NOTHING_TO_INLINE")
+inline fun <A> MaybeKOf<A>.fix(): MaybeK<A> =
+  this as MaybeK<A>
+
+fun <A> Maybe<A>.k(): MaybeK<A> = MaybeK(this)
+
+@Suppress("UNCHECKED_CAST")
+fun <A> MaybeKOf<A>.value(): Maybe<A> = fix().maybe as Maybe<A>
+
+data class MaybeK<out A>(val maybe: Maybe<A>) : MaybeKOf<A> {
+
+  suspend fun suspended(): A? = suspendCoroutine { cont ->
+    value().subscribe(onSuccess = cont::resume, onError = cont::resumeWithException, onComplete = { cont.resume(null) })
+  }
+
+  fun <B> map(f: (A) -> B): MaybeK<B> =
+    maybe.map(f).k()
+
+  fun <B> ap(fa: MaybeKOf<(A) -> B>): MaybeK<B> =
+    flatMap { a -> fa.fix().map { ff -> ff(a) } }
+
+  fun <B> flatMap(f: (A) -> MaybeKOf<B>): MaybeK<B> =
+    maybe.flatMap { f(it).value() }.k()
+
+  /**
+   * A way to safely acquire a resource and release in the face of errors and cancellation.
+   * It uses [ExitCase] to distinguish between different exit cases when releasing the acquired resource.
+   *
+   * @param use is the action to consume the resource and produce an [MaybeK] with the result.
+   * Once the resulting [MaybeK] terminates, either successfully, error or disposed,
+   * the [release] function will run to clean up the resources.
+   *
+   * @param release the allocated resource after the resulting [MaybeK] of [use] is terminates.
+   *
+   * ```kotlin:ank:playground
+   * import arrow.fx.*
+   * import arrow.fx.reaktive.*
+   * import arrow.fx.typeclasses.ExitCase
+   *
+   * class File(url: String) {
+   *   fun open(): File = this
+   *   fun close(): Unit {}
+   *   fun content(): MaybeK<String> =
+   *     MaybeK.just("This file contains some interesting content!")
+   * }
+   *
+   * fun openFile(uri: String): MaybeK<File> = MaybeK { File(uri).open() }
+   * fun closeFile(file: File): MaybeK<Unit> = MaybeK { file.close() }
+   *
+   * fun main(args: Array<String>) {
+   *   //sampleStart
+   *   val safeComputation = openFile("data.json").bracketCase(
+   *     release = { file, exitCase ->
+   *       when (exitCase) {
+   *         is ExitCase.Completed -> { /* do something */ }
+   *         is ExitCase.Canceled -> { /* do something */ }
+   *         is ExitCase.Error -> { /* do something */ }
+   *       }
+   *       closeFile(file)
+   *     },
+   *     use = { file -> file.content() }
+   *   )
+   *   //sampleEnd
+   *   println_atomic(safeComputation)
+   * }
+   *  ```
+   */
+
+  fun <B> bracketCase(use: (A) -> MaybeKOf<B>, release: (A, ExitCase<Throwable>) -> MaybeKOf<Unit>): MaybeK<B> =
+    value()
+      .flatMap { value: A ->
```

Comment:
This is not a correct bracket implementation. I just noticed that the documentation can use some additional clarification here... The idea behind `bracket` is to compose `Maybe` (or `F`), in a resource-safe way
So compose an `acquire: Maybe<A>`, `use: (A) -> Maybe<B>` and `release: (A, ExitCase) -> Maybe<Unit>` in a resource-safe way.

This is only possible, if we can  _guarantee_ that `release` runs _uncancelable_ after`acquire`.

In the snippet above you can cancel between `value()` (acquire) and `use/release` due to the `flatMap` boundary, and thus there is no guarantee that `release` will be called upon cancellation.

Additionally, we also have to make `acquire` uncancelable or the resource could still be leaked with for example `Maybe.fromCallable { resource }.flatMap { println("acquired") }`.

So `bracketCase` runs in two steps.

1. Runs `acquire` in an uncancelable manner
2. Runs `use` in a cancelable manner, while guarantee-ing `release` to be called.

You can also see that reflected in the tests mentioned in your PR in [arrow-test](https://github.com/arrow-kt/arrow-test/pull/10)

- `acquireBracketIsNotCancelable` tests that the `acquire` step cannot be cancelled.
- `releaseBracketIsNotCancelable` tests that `release` cannot be cancelled.
- `cancelOnBracketReleases` tests that `release` is called on `cancel` of use or `acquire`.
   => Additional note on last law, if you cancel during `acquire` you should immediately call `release` after `acquire` and skip `use`. That's the behavior of immediately cancelling a resource-safe program.

Reply:
Some additional information in https://github.com/arrow-kt/arrow-test/pull/10